src/rdd/rdd.rs

```diff
@@ -611,11 +611,13 @@ pub trait Rdd: RddBase + 'static {
                 move |index: usize, partition: Box<dyn Iterator<Item = Self::Item>>|
                 -> Box<dyn Iterator<Item = Self::Item>>
                 {
-                    let bcs = BernoulliCellSampler::new(lower_bound, upper_bound, false);
-                    let new_seed = seed_val + index as u64;
-                    let mut rng = utils::random::get_default_rng_from_seed(new_seed);
+                    let bcs =
+                        Arc::new(BernoulliCellSampler::new(lower_bound, upper_bound, false))
+                        as Arc<dyn RandomSampler<Self::Item>>;
 
-                    Box::new(partition.filter(move |_x: &Self::Item| bcs.sample(&mut rng) > 0))
+                    let sampler_func = bcs.get_sampler(Some(seed_val + index as u64));
+
+                    Box::new(sampler_func(partition).into_iter())
```

Comment:
@iduartgomez Thanks for the hint, I refactored it as you suggested and it is working now.